### PR TITLE
[FLINK-13198] Introduce TimeUtils with parsing method

### PR DESCRIPTION
**Pull request by TsReaper**
## What is the purpose of the change

We need a parse function to parse strings like "123ms", "321s", "12min" and such, so that users can write configurations like "cache.ttl = 1min" and the user input can be changed into a Java `Duration`.

This pull request introduce `TimeUtils` with `parseDuration` method to meet the above needs.

## Brief change log

  - Introduce `TimeUtils`

## Verifying this change

This change can be covered by new tests (`TimeUtilsTest`).

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with @Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-core/src/test/java/org/apache/flink/configuration/TimeLengthTest.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+		assertEquals(1234, TimeLength.parse("1234").getMilliseconds());
+		assertEquals(1234, TimeLength.parse("1234ms").getMilliseconds());
+		assertEquals(1234, TimeLength.parse("1234 ms").getMilliseconds());
+	}
+
+	@Test
+	public void testParseSeconds() {
+		assertEquals(667766, TimeLength.parse("667766s").getSeconds());
+		assertEquals(667766, TimeLength.parse("667766 s").getSeconds());
+		assertEquals(667766, TimeLength.parse("667766s").getSeconds());
```
Comment (wuchong) [inline flink-core/src/test/java/org/apache/flink/configuration/TimeLengthTest.java]: The same with L94?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/DescriptorProperties.java [hunk 2]
```diff
@@ -540,13 +541,33 @@ public TableSchema getTableSchema(String key) {
...
 		return getOptionalMemorySize(key).orElseThrow(exceptionSupplier(key));
 	}
 
+	/**
+	 * Returns a Flink {@link TimeLength} under the given existing key.
+	 */
+	public TimeLength getTimeLength(String key) {
+		return getOptionalTimeLength(key).orElseThrow(exceptionSupplier(key));
+	}
+
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/DescriptorProperties.java]: Please also add `validateTimeLength` similar to `validateMemorySize`.

File: flink-core/src/test/java/org/apache/flink/configuration/TimeLengthTest.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+	}
+
+	@Test
+	public void testParseMinutes() {
+		assertEquals(7657623, TimeLength.parse("7657623min").getMinutes());
+		assertEquals(7657623, TimeLength.parse("7657623 min").getMinutes());
+		assertEquals(7657623, TimeLength.parse("7657623min").getMinutes());
```
Comment (wuchong) [inline flink-core/src/test/java/org/apache/flink/configuration/TimeLengthTest.java]: The same with L101?

File: flink-core/src/test/java/org/apache/flink/configuration/TimeLengthTest.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+	}
+
+	@Test
+	public void testPraseHours() {
+		assertEquals(987654, TimeLength.parse("987654h").getHours());
+		assertEquals(987654, TimeLength.parse("987654 h").getHours());
+		assertEquals(987654, TimeLength.parse("987654h").getHours());
```
Comment (wuchong) [inline flink-core/src/test/java/org/apache/flink/configuration/TimeLengthTest.java]: The same with L108?

**Comment by wuchong** [comment]
cc @zentol , I think you may also want to have a look at this?

**Comment by wuchong** [comment]
Thanks @TsReaper . 

LGTM.

**Comment by wuchong** [comment]
Thanks @TsReaper for the quick updating.

**Comment by TsReaper** [comment]
Thanks @twalthr for the comment. I think `org.apache.flink.api.common.time.Time` has a different functionality compared to the newly introduced `TimeInterval`, so it's ok to keep both the two classes.

**Comment by zentol** [comment]
I'D like to get to the drawing board for this one; it seems this could be trivially solved using Java 8 durations.

**Comment by wuchong** [comment]
Hi @zentol , do you mean java 8 `Duration.parse`? AFAIK, `Duration.parse` only accept ISO-8601 duration format, i.e. `PnDTnHnMn.nS`. For example: `Duration.parse("2s")` is invalid, and `Duration.parse("PT2S")` is valid, however, this is not what we want.

**Comment by TsReaper** [comment]
After discussing with @wuchong and @zentol , we decide not to introduce a new `TimeInterval` class but to add a `TimeUtils` util class with a parsing method.

I've uploaded the commit. Please take a look @wuchong @zentol .

**Comment by wuchong** [comment]
LGTM. +1

**Comment by wuchong** [comment]
Merged.

**Pull request closed without merging**